### PR TITLE
fix(42099): Declaration emit from javascript doesn't respect optional properties in function parameters

### LINKED ISSUE #42099: Declaration emit from javascript doesn't respect optional properties in function parameters
# Bug Report

### 🔎 Search Terms

* javascript declaration
* optional object parameters

### 🕗 Version & Regression Information

* typescript 4.0
* typescript 4.1
* typescript@beta

- This changed between versions 3.x and 4.x

### ⏯ Playground Link

[Playground link with relevant code (check .D.TS tab)](https://www.typescriptlang.org/play?target=7&useJavaScript=true#code/PQKhCgAIUgXBTAzrKFqQAIAcCuAjAGwEsBjVTLAQwCdKBbSAbwHs8AreE2AX0ma1iJy2GvSYA7HHTzxqvfoIB0lYVVoNGk6bN4BtBYkV4AuqtEatMuZH0DDJU+gzV4sHNXGIJUq91TBwADMccS4iZnE4JFgACkZKABpIPCSSbgBKJihIFzcPSEpIAGpk4sgSAG5wP3ASCMRmAnhFAmYAcxiEZDjKAC5IACYM9IqgA)

### 💻 Code

<details>
<summary>jsconfig.json</summary>

```json
{
  "compilerOptions": {
    "module": "es6",
    "moduleResolution": "node",
    "target": "es2020",
    "checkJs": true,
    "declaration": true,
    "noEmit": false,
    "emitDeclarationOnly": true
  }
}
```
</details>

<details>
<summary>test.js</summary>

```javascript
/**
 * test
 *
 * @public
 * @param {object} opts
 * @param {number} opts.a
 * @param {number} [opts.b]
 * @param {number} [opts.c]
 * @returns {number}
 */
function test({a, b, c}) {
  return a + b + c;
}
```
</details>

```
npx tsc  -p jsconfig.json
```

### 🙁 Actual behavior

The definition file doesn't mark b and c as optional

<details>
<summary>test.d.ts</summary>

```ts
/**
 * test
 *
 * @public
 * @param {object} opts
 * @param {number} opts.a
 * @param {number} [opts.b]
 * @param {number} [opts.c]
 * @returns {number}
 */
declare function test({ a, b, c }: {
    a: number;
    b: number;
    c: number;
}): number;
```
</details>

### 🙂 Expected behavior

The definition file should mark b and c as optional (so that projects importing the definitions will be checked properly)

<details>
<summary>test.d.ts</summary>

```ts
/**
 * test
 *
 * @public
 * @param {object} opts
 * @param {number} opts.a
 * @param {number} [opts.b]
 * @param {number} [opts.c]
 * @returns {number}
 */
declare function test({ a, b, c }: {
    a: number;
    b?: number;
    c?: number;
}): number;
```
</details>

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #42099

### CODE REVIEW

File: tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * @param {object} opts
+ * @param {number} opts.a
+ * @param {number} [opts.b]
+ * @param {number} [opts.c]
+ * @returns {number}
+ */
+declare function foo({ a, b, c }: {
+    a: number;
+    b?: number;
+    c?: number;
```
Comment (andrewbranch) [inline tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js:47]: These should be `number | undefined` for consistency with emit elsewhere.

**Review by andrewbranch — CHANGES_REQUESTED** [review]

File: tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js (hunk 1 above)
Comment (a-tarasyuk) [reply tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js:47]: @andrewbranch Do we need to follow the same behavior as in the TypeScript example?


```ts
function foo(args: { x: number; y?: number }) {
}

//  .d.ts

declare function foo(args: {
    x: number;
    y?: number;
}): void;
```

[Playground](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUBDATgcwM4BciA3ogB5FggC2ARgKZYDciAngPyU0NaIC+AShIAoPsKA)

Comment (a-tarasyuk) [reply tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js:47]: Or `| undefined` should be added only for JSDoc?

Comment (andrewbranch) [reply tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js:47]: Oh hang on—this test doesn’t have `--strictNullChecks`, my bad. I bet it already works for `--strictNullChecks`. I’m not sure whether the TS example you just gave is a bug. It probably reuses the type annotation node provided. But _generally_, when we serialize optional properties, we add the `| undefined` to the type annotation even if it wasn’t written by the user. But that would never show up without `--strictNullChecks`. Can we enable it for this test just to make sure we know how it behaves?

Comment (a-tarasyuk) [reply tests/baselines/reference/jsDeclarationsOptionalTypeLiteralProps.js]: @andrewbranch Sure! I've added [jsDeclarationsOptionalTypeLiteralProps2](https://github.com/microsoft/TypeScript/pull/42116/files#diff-208d846f4d46557112e779f86ad593ae2ee869042212e5e2ff564a6edf8164dc) with the `strict` option.

**Review by andrewbranch — APPROVED** [review]
Thanks @a-tarasyuk!

**Pull request merged**